In [4]:
import os
print("Current folder:", os.getcwd())

Current folder: C:\Users\DELL\Documents\telecom-churn-analytics\notebooks


In [6]:
import pandas as pd

# Load dataset
df = pd.read_csv('C:/Users/DELL/Documents/telecom-churn-analytics/data/raw/WA_Fn-UseC_-Telco-Customer-Churn.csv')

# Verify
print(f"✓ Rows    : {df.shape[0]}")
print(f"✓ Columns : {df.shape[1]}")
print(f"✓ Churn counts:\n{df['Churn'].value_counts()}")

✓ Rows    : 7043
✓ Columns : 21
✓ Churn counts:
Churn
No     5174
Yes    1869
Name: count, dtype: int64


# EDA: IBM Telco Customer Churn Analysis
**Dataset:** 7,043 customers · 21 columns  
**Source:** Kaggle — blastchar/telco-customer-churn  
**Objective:** Identify which customer segments are most at risk of churning  
**Tools:** Python, Pandas, Matplotlib, Seaborn

In [7]:
# Import all libraries needed for EDA
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Display settings — makes output cleaner
pd.set_option('display.max_columns', 30)
pd.set_option('display.float_format', '{:.2f}'.format)
plt.style.use('seaborn-v0_8-whitegrid')

print("✓ All libraries imported successfully")

✓ All libraries imported successfully


In [8]:
# Load the raw dataset using relative path
df = pd.read_csv('C:/Users/DELL/Documents/telecom-churn-analytics/data/raw/WA_Fn-UseC_-Telco-Customer-Churn.csv')

print(f"✓ Dataset loaded successfully")
print(f"✓ Rows    : {df.shape[0]:,}")
print(f"✓ Columns : {df.shape[1]}")

✓ Dataset loaded successfully
✓ Rows    : 7,043
✓ Columns : 21


In [9]:
# See all columns with their data types
print("Column Name          | Data Type")
print("-" * 40)
for col in df.columns:
    print(f"{col:<25} {df[col].dtype}")

Column Name          | Data Type
----------------------------------------
customerID                object
gender                    object
SeniorCitizen             int64
Partner                   object
Dependents                object
tenure                    int64
PhoneService              object
MultipleLines             object
InternetService           object
OnlineSecurity            object
OnlineBackup              object
DeviceProtection          object
TechSupport               object
StreamingTV               object
StreamingMovies           object
Contract                  object
PaperlessBilling          object
PaymentMethod             object
MonthlyCharges            float64
TotalCharges              object
Churn                     object


## Data Dictionary

| Column | Type | Meaning |
|---|---|---|
| customerID | Text | Unique customer identifier |
| gender | Text | Male / Female |
| SeniorCitizen | 0/1 | Whether customer is 65+ years old |
| Partner | Yes/No | Whether customer has a partner |
| Dependents | Yes/No | Whether customer has dependents |
| tenure | Number | Months the customer has been with the company |
| PhoneService | Yes/No | Has phone service |
| MultipleLines | Yes/No | Has multiple phone lines |
| InternetService | Text | DSL / Fiber optic / No |
| OnlineSecurity | Yes/No | Has online security add-on |
| OnlineBackup | Yes/No | Has online backup add-on |
| DeviceProtection | Yes/No | Has device protection add-on |
| TechSupport | Yes/No | Has tech support add-on |
| StreamingTV | Yes/No | Streams TV |
| StreamingMovies | Yes/No | Streams movies |
| Contract | Text | Month-to-month / One year / Two year |
| PaperlessBilling | Yes/No | Uses paperless billing |
| PaymentMethod | Text | Electronic check / Mailed check / Bank transfer / Credit card |
| MonthlyCharges | Number | Current monthly bill amount |
| TotalCharges | Number | Total amount charged (STORED AS TEXT — bug) |
| Churn | Yes/No | Whether customer left the company ← TARGET VARIABLE |

In [10]:
# Look at the first 5 rows
df.head()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,OnlineBackup,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,Yes,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,No,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,Yes,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,No,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,No,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


In [11]:
# Last 5 rows — checks if data is consistent end-to-end
df.tail()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,OnlineBackup,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
7038,6840-RESVB,Male,0,Yes,Yes,24,Yes,Yes,DSL,Yes,No,Yes,Yes,Yes,Yes,One year,Yes,Mailed check,84.80,1990.5,No
7039,2234-XADUH,Female,0,Yes,Yes,72,Yes,Yes,Fiber optic,No,Yes,Yes,No,Yes,Yes,One year,Yes,Credit card (automatic),103.20,7362.9,No
7040,4801-JZAZL,Female,0,Yes,Yes,11,No,No phone service,DSL,Yes,No,No,No,No,No,Month-to-month,Yes,Electronic check,29.60,346.45,No
7041,8361-LTMKD,Male,1,Yes,No,4,Yes,Yes,Fiber optic,No,No,No,No,No,No,Month-to-month,Yes,Mailed check,74.40,306.6,Yes
7042,3186-AJIEK,Male,0,No,No,66,Yes,No,Fiber optic,Yes,No,Yes,Yes,Yes,Yes,Two year,Yes,Bank transfer (automatic),105.65,6844.5,No


In [12]:
# Random 5 rows — gives unbiased view of the data
df.sample(5, random_state=42)

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,OnlineBackup,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
185,1024-GUALD,Female,0,Yes,No,1,No,No phone service,DSL,No,No,No,No,No,No,Month-to-month,Yes,Electronic check,24.80,24.8,Yes
2715,0484-JPBRU,Male,0,No,No,41,Yes,Yes,No,No internet service,No internet service,No internet service,No internet service,No internet service,No internet service,Month-to-month,Yes,Bank transfer (automatic),25.25,996.45,No
3825,3620-EHIMZ,Female,0,Yes,Yes,52,Yes,No,No,No internet service,No internet service,No internet service,No internet service,No internet service,No internet service,Two year,No,Mailed check,19.35,1031.7,No
1807,6910-HADCM,Female,0,No,No,1,Yes,No,Fiber optic,No,No,Yes,No,No,No,Month-to-month,No,Electronic check,76.35,76.35,Yes
132,8587-XYZSF,Male,0,No,No,67,Yes,No,DSL,No,No,No,Yes,No,No,Two year,No,Bank transfer (automatic),50.55,3260.1,No


In [13]:
# Statistical summary of numeric columns
df.describe()

,SeniorCitizen,tenure,MonthlyCharges
count,7043.00,7043.00,7043.00
mean,0.16,32.37,64.76
std,0.37,24.56,30.09
min,0.00,0.00,18.25
25%,0.00,9.00,35.50
50%,0.00,29.00,70.35
75%,0.00,55.00,89.85
max,1.00,72.00,118.75


### Observations from describe()
- **tenure** ranges from 0 to 72 months (6 years max). Average is 32 months.
- **MonthlyCharges** ranges from $18 to $119. Average is $64.76.
- **SeniorCitizen** mean is 0.16 — meaning only 16% of customers are seniors.
- TotalCharges is missing from this table because it's stored as text (object) — data quality issue to fix.

In [15]:
# Check missing values in every column
missing = df.isnull().sum()
missing_pct = (df.isnull().sum() / len(df)) * 100

missing_df = pd.DataFrame({
    'Missing Count': missing,
    'Missing %': missing_pct
})

# Only show columns that have missing values
print("Columns with missing values:")
print(missing_df[missing_df['Missing Count'] > 0])
print("\nColumns with NO missing values:", (missing_df['Missing Count'] == 0).sum())

Columns with missing values:
Empty DataFrame
Columns: [Missing Count, Missing %]
Index: []

Columns with NO missing values: 21


In [16]:
# Check for duplicate customers
duplicates = df.duplicated().sum()
print(f"Duplicate rows: {duplicates}")

# Also check if customerID is unique
unique_ids = df['customerID'].nunique()
print(f"Unique customer IDs: {unique_ids}")
print(f"Total rows: {len(df)}")
print(f"IDs match rows: {unique_ids == len(df)}")

Duplicate rows: 0
Unique customer IDs: 7043
Total rows: 7043
IDs match rows: True


In [18]:
# How many customers churned?
churn_counts = df['Churn'].value_counts()
churn_pct = df['Churn'].value_counts(normalize=True) * 100

print("Churn Distribution:")
print(f"  Stayed  (No)  : {churn_counts['No']:,}  ({churn_pct['No']:.1f}%)")
print(f"  Churned (Yes) : {churn_counts['Yes']:,}  ({churn_pct['Yes']:.1f}%)")
print(f"\nFor every 1 churned customer, {churn_counts['No']/churn_counts['Yes']:.1f} stayed")

Churn Distribution:
  Stayed  (No)  : 5,174  (73.5%)
  Churned (Yes) : 1,869  (26.5%)

For every 1 churned customer, 2.8 stayed


### Key Business Insight #1
**26.5% churn rate** — roughly 1 in 4 customers leaves.  
From a base of 7,043 customers, **1,869 have churned**.  
This is the class imbalance we must handle in the ML model on Day 3.  
If MonthlyCharges average is $64, churned customers represent  
**~$120,000/month in lost revenue**.